KOGITO-7648: KIE Tools - Integration tests intermittent faillure

domhanak opened the pull request:
Following up on your PR #1143 will add stabilization here, have some ideas.

Starting with the upgrade, will follow up with split of the test cases and hopefully there will be stability.

### 
**List of runs & status**

1. Build1 -`@kie-tools/boxed-expression-component@0.0.0 cy:run` -> ` Context Expression Tests :: Nested Relations` -> `Keyboard interaction with header's contextMenu in nested decision table` - Test failed to locate element. I do not see any correlation with changes in this PR. :yellow_circle: 
2. Build2 - Second run was successful :green_circle: 
3. Build3 -`@kie-tools/boxed-expression-component@0.0.0 cy:run` -> ` Context Expression Tests :: Nested Relations` -> `Keyboard interaction with header's contextMenu in nested decision table` :yellow_circle: 
4. Build4 - Success :green_circle: 
5. Build5 - Success :green_circle: 
6. Build6 - `@kie-tools/boxed-expression-component@0.0.0 cy:run` -> ` Context Expression Tests :: Nested Relations` -> `Keyboard interaction with header's contextMenu in nested decision table` :yellow_circle:
7. Build7 - Success :green_circle: 
8. Build8 - `chrome-extension-pack-kogito-kie-editors@0.0.0 build:prod` -> `BpmnTest :: StaleElementReferenceError: stale element reference: element is not attached to the page document` :yellow_circle: 
9. Build9 - `vscode-extension-pack-kogito-kie-editors@0.0.0 build:prod` -> `Process with UserTask.bpmn test` :: `StaleElementReferenceErrorstale element reference: element is not attached to the page document` :yellow_circle: 
10. Build10 - Success :green_circle: 
11. Build11 - `chrome-extension-pack-kogito-kie-editors@0.0.0 build:prod` -> `DmnExpressionEditorTest :: StaleElementReferenceError: stale element reference: element is not attached to the page document` :yellow_circle: 
12. Build12 - Success :green_circle:

yesamer commented:
Jenkins retest this

yesamer commented:
jenkins retest this please

domhanak commented:
@yesamer Any ideas about the `-boxed-expression-component` intermittent failures?

domhanak commented:
@yesamer the jenkins job works but it is not being triggered

jenkins test this

yesamer commented:
No, maybe @fantonangeli can help?

domhanak commented:
Added some more fixes, turns out there was a small issue when there was a failure. Added some code to fix it.

@yesamer @fantonangeli shall I disable the test? it looks like it s more unstable then the vscode

fantonangeli commented:
@domhanak I think you can disable the tests, not to blocks others work, and next week when I come back from holiday I can look at it.
In this case, will you please open a Jira?

domhanak commented:
@fantonangeli Ack, creating a ticket and will disable when the builds finish :) Thanks! Enjoy :palm_tree:

domhanak commented:
Skipped and reported  https://issues.redhat.com/browse/KOGITO-7790

@yesamer Wdyt, shall we continue with merging or do we try more?

yesamer commented on packages/vscode-extension-pack-kogito-kie-editors/package.json:
```diff
@@ -23,11 +23,11 @@
     "run:webmode": "pnpm vscode-test-web --browserType=chromium --extensionDevelopmentPath=. --version=stable",
     "test": "pnpm run-script-if --bool \"$(build-env global.build.test)\" --then \"jest --silent --verbose --passWithNoTests\"",
-    "test:it": "pnpm run-script-if --bool \"$(build-env global.build.testIT)\" --then \"pnpm test:it:clean\" \"cpr it-tests/resources it-tests-tmp/resources\" \"tsc --project tsconfig.it-tests.json --skipLibCheck --sourceMap true\" \"extest setup-and-run --yarn -c 1.66.2 -u -e ./test-resources -o it-tests/settings.json out/*.test.js\"",
+    "test:it": "pnpm run-script-if --bool \"$(build-env global.build.testIT)\" --then \"pnpm test:it:clean\" \"cpr it-tests/resources it-tests-tmp/resources\" \"tsc --project tsconfig.it-tests.json --skipLibCheck --sourceMap true\" \"extest setup-and-run --yarn -u -e ./test-resources -o it-tests/settings.json out/*.test.js\"",
```
@domhanak I think removing the VSCode version could introduce instability. `vscode-extension-tester` supports VSCode version 1.69 ([as described here](https://github.com/redhat-developer/vscode-extension-tester/blob/master/CHANGELOG.md)). Removing the  `-c` parameter with the precise VSCode version to use, it means using the latest VSCode version (currently 1.70).
Thus, if we don't frequently update `vscode-extension-tester`, this can result in instability, mainly because VSCode evolves very quickly.
Said so, we have two solutions here:
- Commit ourselves to keep `vscode-extension-tester` always updated.
- Keep the `-c` version with the last supported VSCode version. 

My choice is the second option, which should avoid any instability due to unsupported (and newer) VSCode versions.
Then, my suggestion is to keep the `-c` option, updating the VSCode version to `1.69` 
WDYT?

domhanak replied:
Well, the compatibility is a bit tricky I would say. If there is no new components the tests will work imho. Only in case some already known VSCode components change their locators will it break.

yesamer replied:
Partially agree with you, we can consider this risk very low, but it can raise with any new `VSCode` version. Don't consider only the new `VSCode` feature, but potential bugs introduced in the new releases. To give you an example, the root cause of the issue I found (https://github.com/redhat-developer/vscode-extension-tester/issues/466) is a bug on `chromedriver`. Depending on the `chomedriver` version used by a specific `VSCode` extension, the issue can be present or not.
The point is that hardcoding the last supported `VSCode` version gives us more stability, even if we don't update for months. 
However, if you prefer to use the stable version (the latest), fine for me to proceed. In this case, my suggestion is to keep `vscode-extension-tester` as updated as possible to reducing that risk.

yesamer replied:
and.. I would use the same approach in all the modules.
So, whatever your choice is, I suggest updating the `test:it` in other modules (`yard`, `swf`) accordingly. 
Thank you!

domhanak replied:
@yesamer what about we introduce a `KOGITO_BUILD_testIT_vscode_version` as a build-end property? Does it make sense?

tiagobento replied:
@domhanak In this case, I think that just adding a new env entry would be enough, no need for an env variable. Env variables are meant to be used when there's the need for changing it during a build targeting a specific goal (e.g. Dev, Staging, Prod).

yesamer replied:
@domhanak That can be a good alternative that we can evaluate in case that risk becomes reality. Ping me when you want to merge this.

domhanak replied:
Fixed the tests to `1.69`

yesamer commented on packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/VSCodeTestHelper.ts:
```diff
@@ -145,12 +145,16 @@ export default class VSCodeTestHelper {
+          await dialog.pushButton("Don't Save");
+        }
+      } catch (noSaveDialogWhenClosingError) {
+        // no-0p
```
@domhanak Can you please explain in which scenario this `noSaveDialogWhenClosingError` can occur? Thx!

domhanak replied:
@yesamer it is another case that a test can be in after execution.  I was able to reproduce by failing some assert in test. The whole flow stopped and I got error that test expected `monaco-dialog-box` (save dialog) introducingw this block allowed the to proceed.

yesamer replied:
@domhanak ok, thank you. Usually, I don't recommend silencing a `catch` block, I would consider just adding a log to trace it. However, consider it an optional suggestion.

domhanak replied:
@yesamer Sure logging it sounds good, will do.

yesamer reviewed (approved):
@domhanak Thank you for improving the `it:tests`!

domhanak commented:
Rebased on main.

yesamer commented:
@domhanak thank you for listening and for drastically improving our tests stability!

domhanak commented:
Used unavailable version, latest commit fixes it to `1.69.2` - if green I think we can proceed @tiagobento @yesamer  let me know

yesamer commented:
2022-08-22T15:04:10.4317562Z [run-script-if] Running 'extest setup-and-run --yarn -c 1.69 -u -e ./test-resources -o it-tests/settings.json out/*.test.js'
2022-08-22T15:04:11.3926362Z Error: Version 1.69 is not available in stable stream
2022-08-22T15:04:11.3927923Z     at CodeUtil.<anonymous> (/home/runner/work/kie-tools/kie-tools/node_modules/.pnpm/vscode-extension-tester@4.4.1_mocha@9.2.0/node_modules/vscode-extension-tester/out/util/codeUtil.js:306:23)
2022-08-22T15:04:11.3928695Z     at Generator.next (<anonymous>)
2022-08-22T15:04:11.3929867Z     at fulfilled (/home/runner/work/kie-tools/kie-tools/node_modules/.pnpm/vscode-extension-tester@4.4.1_mocha@9.2.0/node_modules/vscode-extension-tester/out/util/codeUtil.js:5:58)
2022-08-22T15:04:11.3930825Z     at processTicksAndRejections (node:internal/process/task_queues:96:5)
2022-08-22T15:04:11.4028298Z [run-script-if] Error executing 'extest setup-and-run --yarn -c 1.69 -u -e ./test-resources -o it-tests/settings.json out/*.test.js'.

@domhanak Investigating on the reason

yesamer commented:
@domhanak Great you already investigated on it :) Thank you

domhanak commented:
@yesamer no worries, looking after my PR :) Looks like an unstable test in `chrome-extension-pack-kogito-kie-editors` now. :bomb:

yesamer commented:
@domhanak Yep, I already saw that failure on `chrome-extension-pack-kogito-kie-editors` in the past, I think that is the next (and hopefully) last step to close this process.

domhanak commented:
@yesamer once this run finished I ll push one more update of dependencies for chrome-extension.
`selenium-webdriver`  to `4.4.0` and its types to `4.1.2` maybe it will help.

fantonangeli commented:
@domhanak, in this PR I fixed the `context_expression_spec` failing tests.
https://github.com/domhanak/kie-tools/pull/23

yesamer commented:
2022-08-23T18:02:04.7661984Z   Serverless workflow editor - smoke integration tests
2022-08-23T18:02:16.8024054Z       ✔ Opens greetings.sw.json and loads two editor groups: 8393ms
2022-08-23T18:02:18.6934525Z Shutting down the browser
2022-08-23T18:02:19.5618759Z Uninstalling redhat.vscode-extension-serverless-workflow-editor...
2022-08-23T18:02:19.7182646Z Extension 'redhat.vscode-extension-serverless-workflow-editor' was successfully uninstalled!
2022-08-23T18:02:19.7583404Z 
2022-08-23T18:02:19.7584112Z   Suite duration: 14.993 s, Tests: 1
2022-08-23T18:02:19.7584390Z 
2022-08-23T18:02:19.7584583Z   1 passing (46s)
2022-08-23T18:02:19.7584881Z   1 failing
2022-08-23T18:02:19.7585074Z 
2022-08-23T18:02:19.7589946Z   1) Serverless workflow editor - autocompletion tests
2022-08-23T18:02:19.7590659Z        Completes serverless workflow with function and state autocompletion:
2022-08-23T18:02:19.7590953Z 
2022-08-23T18:02:19.7591258Z       AssertionError: expected 0 to equal 3
2022-08-23T18:02:19.7591692Z       + expected - actual
2022-08-23T18:02:19.7591905Z 
2022-08-23T18:02:19.7592080Z       -0
2022-08-23T18:02:19.7592360Z       +3
2022-08-23T18:02:19.7592622Z       
2022-08-23T18:02:19.7593462Z       at Context.<anonymous> (out/serverless-workflow-editor-extension-autocompletion.test.js:195:59)
2022-08-23T18:02:19.7594405Z       at step (out/serverless-workflow-editor-extension-autocompletion.test.js:33:23)
2022-08-23T18:02:19.7595358Z       at Object.next (out/serverless-workflow-editor-extension-autocompletion.test.js:14:53)
2022-08-23T18:02:19.7596282Z       at fulfilled (out/serverless-workflow-editor-extension-autocompletion.test.js:5:58)
2022-08-23T18:02:19.7596840Z       at runMicrotasks (<anonymous>)
2022-08-23T18:02:19.7597423Z       at processTicksAndRejections (node:internal/process/task_queues:96:5)

domhanak commented:
@yesamer Looks like an issue with the autocompletion test. Checking.

domhanak commented:
Added the chrome-extension dep upgrade & investigating the failures of swf locally.

yesamer commented on pnpm-lock.yaml:
```diff
@@ -21929,6 +21947,19 @@ packages:
       - utf-8-validate
     dev: true
 
+  /selenium-webdriver/4.4.0:
```
@domhanak Is `/selenium-webdriver/4.2.0` still used in other projects? If not, I suggest to remove it and using version `4.4.0` only

yesamer replied:
ok - you reverted it - nevermind :)

domhanak replied:
Yea, I will create a JIRA to upgrade selenium at once for all in separate PR, think I mix too much here already.

tiagobento commented:
Hey guys, thanks for so much for this work. I'm glad to see you all working together towards making our test suites more stable 🎉

Since you're all at it, I just have a small comment regarding Chrome Extension tests. Although they have been amazingly reliable, I think they are the ones with the highest risk of failing over time, given their need of connecting directly with GitHub.com, which is something we don't control. For this reason, maybe they could benefit from a retry configuration like the one Dominik added to VS Code Extension tests. WDYT?

yesamer commented:
@tiagobento +1
@domhanak Can you please merge Fabrizio's fix? After that, I think we can finalize this PR (if we have all green checks), and concentrate on the `chrome-extensions-test` as the next step in another ticket.

domhanak commented:
@yesamer Fabrizio's fix is already merged. Let' see the build.

yesamer commented:
@domhanak He provided 2 PRs. I see only the first merged https://github.com/domhanak/kie-tools/pull/22, the second one https://github.com/domhanak/kie-tools/pull/23 still not merged.

domhanak commented:
My bad, did not notice, merged!

domhanak commented:
@yesamer @tiagobento Ticket for `chrome-extension-pack-kogito-kie-editors` -> https://issues.redhat.com/browse/KOGITO-7845

domhanak commented:
@tiagobento @yesamer Selenium-webdriver upgrade -> https://issues.redhat.com/browse/KOGITO-7846

domhanak commented:
@fantonangeli Hi could you please send the [23](https://github.com/domhanak/kie-tools/pull/23/commits/20d47cf16f9dacc4d8e8f1199cd07c9c67e06629) PR again? I forgot to pull it after latest rebase and it got messed up.

domhanak commented:
Ahhh, rebased again. Missing one Commit for @fantonangeli let me know I can add it somehow but we will lose the history.
Also fixed one missing reassignment that could be the cause of the latest error.

domhanak commented:
@yesamer not sure what the issue is. any idea?

yesamer commented:
@domhanak No :( . Can you please try to rebase it with main?

domhanak commented:
```
FAIL it-tests/tests/DmnExpressionEditorTest.ts
2022-08-26T10:50:29.8431658Z   ● Console
2022-08-26T10:50:29.8431938Z 
2022-08-26T10:50:29.8432092Z     console.error
2022-08-26T10:50:29.8432539Z       Error: Error while clicking on element.
2022-08-26T10:50:29.8434438Z           at Function.<anonymous> (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/utils/tools/ErrorProcessor.ts:28:32)
2022-08-26T10:50:29.8437317Z           at step (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/utils/tools/ErrorProcessor.ts:48:23)
2022-08-26T10:50:29.8438875Z           at Object.next (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/utils/tools/ErrorProcessor.ts:29:53)
2022-08-26T10:50:29.8440370Z           at /home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/utils/tools/ErrorProcessor.ts:23:71
2022-08-26T10:50:29.8441002Z           at new Promise (<anonymous>)
2022-08-26T10:50:29.8443208Z           at Object.<anonymous>.__awaiter (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/utils/tools/ErrorProcessor.ts:19:12)
2022-08-26T10:50:29.8445363Z           at Function.Object.<anonymous>.ErrorProcessor.run (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/utils/tools/ErrorProcessor.ts:66:16)
2022-08-26T10:50:29.8447268Z           at Element.<anonymous> (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/Element.ts:58:33)
2022-08-26T10:50:29.8449157Z           at step (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/Element.ts:48:23)
2022-08-26T10:50:29.8450699Z           at Object.next (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/Element.ts:29:53)
2022-08-26T10:50:29.8452047Z           at /home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/Element.ts:23:71
2022-08-26T10:50:29.8452963Z           at new Promise (<anonymous>)
2022-08-26T10:50:29.8454374Z           at Object.<anonymous>.__awaiter (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/Element.ts:19:12)
2022-08-26T10:50:29.8456623Z           at Element.Object.<anonymous>.Element.click (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/Element.ts:137:16)
2022-08-26T10:50:29.8458683Z           at GitHubListItem.<anonymous> (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/github-file-list/GitHubListItem.ts:31:16)
2022-08-26T10:50:29.8460639Z           at step (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/github-file-list/GitHubListItem.ts:63:23)
2022-08-26T10:50:29.8462348Z           at Object.next (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/github-file-list/GitHubListItem.ts:44:53)
2022-08-26T10:50:29.8464187Z           at fulfilled (/home/runner/work/kie-tools/kie-tools/packages/chrome-extension-pack-kogito-kie-editors/it-tests/framework/github-file-list/GitHubListItem.ts:35:58)
2022-08-26T10:50:29.8465043Z           at processTicksAndRejections (node:internal/process/task_queues:96:5)
2022-08-26T10:50:29.8465289Z 
2022-08-26T10:50:29.8465474Z       33 |     } catch (err) {
2022-08-26T10:50:29.8465857Z       34 |       // print detailed stack trace
2022-08-26T10:50:29.8466262Z     > 35 |       console.error(customError.stack);
2022-08-26T10:50:29.8466580Z          |               ^
2022-08-26T10:50:29.8466858Z       36 |       throw err;
2022-08-26T10:50:29.8467294Z       37 |     }
2022-08-26T10:50:29.8467894Z       38 |   }
2022-08-26T10:50:29.8468029Z 
2022-08-26T10:50:29.8468550Z       at Function.<anonymous> (it-tests/utils/tools/ErrorProcessor.ts:35:15)
2022-08-26T10:50:29.8469398Z       at step (it-tests/utils/tools/ErrorProcessor.ts:48:23)
2022-08-26T10:50:29.8470081Z       at Object.throw (it-tests/utils/tools/ErrorProcessor.ts:29:53)
2022-08-26T10:50:29.8470734Z       at rejected (it-tests/utils/tools/ErrorProcessor.ts:21:65)
2022-08-26T10:50:29.8470962Z 
2022-08-26T10:50:29.8471522Z   ● DmnExpressionEditorTest
2022-08-26T10:50:29.8471693Z 
2022-08-26T10:50:29.8472294Z     StaleElementReferenceError: stale element reference: element is not attached to the page document
2022-08-26T10:50:29.8472775Z       (Session info: chrome=104.0.5112.101)
2022-08-26T10:50:29.8472939Z 
2022-08-26T10:50:29.8473673Z       at Object.throwDecodedError (../../node_modules/.pnpm/selenium-webdriver@4.2.0/node_modules/selenium-webdriver/lib/error.js:522:15)
2022-08-26T10:50:29.8474671Z       at parseHttpResponse (../../node_modules/.pnpm/selenium-webdriver@4.2.0/node_modules/selenium-webdriver/lib/http.js:549:13)
2022-08-26T10:50:29.8475782Z       at Executor.execute (../../node_modules/.pnpm/selenium-webdriver@4.2.0/node_modules/selenium-webdriver/lib/http.js:475:28)
2022-08-26T10:50:29.8477253Z       at Driver.execute (../../node_modules/.pnpm/selenium-webdriver@4.2.0/node_modules/selenium-webdriver/lib/webdriver.js:735:17)
```

domhanak commented:
@tiagobento @yesamer All done imho, let's tackle the chrome-extension in follow-up ticket.
